### ISSUE TITLE
Learning rate is not right in FasterRCNN

### ISSUE THREAD

**Issue by liuzili97** [tensorpack/tensorpack#1081]
I feel there is a bug exists in `ScheduledHyperParamSetter`. The learning_rate is not right.

### 1. What you did:

I'm trying to train a FasterRCNN model. I load a pre-trained model (global step is 11000) and I'm going to continue the training process. I set the `starting_epoch=201` in the TrainConfig.

### 2. What you observed:

After starting training, I found that the learning_rate is 0.01. However, **the learning_rate suppose to be 0.001**, as you can see in the following:
```

[32m[0213 21:22:55 @train.py:348][0m Warm Up Schedule (steps, value): [(0, 0.0033000000000000004), (1000, 0.01)]
[32m[0213 21:22:55 @train.py:349][0m LR Schedule (epochs, value): [(2, 0.01), (120.0, 0.001), (208.0, 0.00010000000000000002)]

^[[32m[0213 21:23:57 @base.py:274]^[[0m Start Epoch 201 ...
^[[32m[0213 21:27:03 @base.py:284]^[[0m Epoch 201 (global_step 110500) finished, time:3 minutes 5 seconds.
^[[32m[0213 21:27:03 @trainers.py:417]^[[0m Running horovod broadcast ...
^[[32m[0213 21:27:03 @misc.py:109]^[[0m Estimated Time Left: 2 hours 59 seconds
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m PeakMemory(MB)/gpu:0: 2849.9
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m QueueInput/queue_size: 50
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m fastrcnn_losses/box_loss: 0.06977
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m fastrcnn_losses/label_loss: 0.085708
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m fastrcnn_losses/label_metrics/accuracy: 0.96229
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m fastrcnn_losses/label_metrics/false_negative: 0.082378
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m fastrcnn_losses/label_metrics/fg_accuracy: 0.91663
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m fastrcnn_losses/num_fg_label: 12.787
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m learning_rate: 0.01
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m rpn_losses/box_loss: 0.0032397
^[[32m[0213 21:27:03 @monitor.py:467]^[[0m rpn_losses/label_loss: 0.00039103

```

I found that the problem above is due to the warm-up schedule in FasterRCNN. The code below is in `ScheduledHyperParamSetter`:
```
def _get_value_to_set(self):
        refnum = self.global_step if self._step else self.epoch_num
        laste, lastv = None, None
        for e, v in self.schedule:
            if e == refnum:
                return v    # meet the exact boundary, return directly
            if e > refnum:
                break
            laste, lastv = e, v
        if laste is None or laste == e:
            # hasn't reached the first scheduled point, or reached the end of all scheduled points
            return None
        if self.interp is None:
            # If no interpolation, nothing to do.
            return None
        v = (refnum - laste) * 1. / (e - laste) * (v - lastv) + lastv
        return v
```
The warm-up schedule is  `(steps, value): [(0, 0.0033000000000000004), (1000, 0.01)]
` and the self.global_step is 11000. So, the learning_rate will stuck in 0.01.

I'm wondering how to solve this problem, could you give me some help? thanks.

### 4. Your environment:
  + Python version: 3.6
  + TF version: 1.12
  + Tensorpack version: 0.9.0.1

**Comment by ppwwyyxx** [comment]
> The warm-up schedule is `(steps, value): [(0, 0.0033000000000000004), (1000, 0.01)] ` and the self.global_step is 11000. So, the learning_rate will stuck in 0.01.

This is not true. As you can see in the code, when the current step is after the end of all scheduled point (1000), it will return `None`.

The second parameter scheduler (the epoch-based one) is epoch base and without interpolation. As you can see in the code above, it is designed to work only at the exact schedule point. 

The learning rate is 0.01 probably because that's learning rate in your checkpoint.

**Comment by ppwwyyxx** [comment]
Related to that, the design choice that's made here can be reconsidered. 
Perhaps it's OK to let the scheduler set the parameter at the beginning of training as well, in addition to at the exact scheduled points.

**Comment by liuzili97** [comment]
> This is not true. As you can see in the code, when the current step is after the end of all scheduled point (1000), it will return `None`, and the parameter will not be changed.

Oh you are right. I didn't notice that `or laste == e` is behind the ` if laste is None`. After I ignore the `learning_rate` when loading the checkpoint, the problem solved. Thanks!

**Comment by ppwwyyxx** [comment]
To avoid future confusion, after e79d74f85, loading a checkpoint with a learning_rate that's inconsistent to the schedule will lead to a warning message like this:
```python
logger.warn("According to scheduler {}, parameter '{}' should become {} at the current point. "
            "However its current value is {}. "
            "If this is the only scheduler being used, you may want to check whether your "
            "initialization of the parameter is as expected".format(
                self, self.param.readable_name, v, actual_value))

```
